src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,594 @@
+        else:
+            lambda_code_text = f"{lambda_code_text}{line}\n"
+    return lambda_code_text.strip()
+
+
+def extract_submodule_imports(
+    func: FunctionType, top_level_modules: Iterable[ModuleType]
+) -> Set[Tuple[str, str]]:
+    """
+    Get submodule imports, the func code co_names only gives the top level module names, the submodule imports
+    have to be inferred manually. Consider the following example:
+
+    import a1.a2.a3.a4
+    def func():
+        a1.a2.a3.a4.foo()
+
+    func.__code__.co_names only contains ("a1", "a2", "a3", "a4", "foo") which does not include the
+    complete import path information.
+
+    To reconstruct "a1.a2.a3.a4", the current strategy is to import each prefix import of the import chains.
+    This is not a perfect solution as we could import modules not used, but it works.
+
+    Args:
+        func: The target function to generate source code for.
+        top_level_modules: The name of top level modules from which to search the referenced imported objects
+            or submodules.
+
+    Returns:
+        A set of tuple with each tuple composed of two string, the first one is actual name for the imported object,
+        and the second one is alias.
+
+    """
+    imports = set()
+    for module in top_level_modules:
+        module_prefix = f"{module.__name__}."
+        for name in [m for m in sys.modules if m.startswith(module_prefix)]:
+            tokens = set(name[len(module_prefix) :].split("."))
+            if not tokens - set(
+                func.__code__.co_names
+            ):  # only add imports that co_names contains
+                imports.add((name, name))
+    return imports
+
+
+def find_target_func_objects_references(
+    func: Union[FunctionType, BuiltinFunctionType],
+    to_import: Set[Tuple[str, str]],
+    ref_objects: Dict[str, Any],
+    classes_to_generate: List[type],
+) -> None:
+    """
+    Find objects referenced by functions including classes, methods, modules, global variables.
+
+    Args:
+        func: The target function to generate source code for.
+        to_import: set of name and alias pairs of direct imports which should be generated as "import xxx"
+            or "import xxx as yyy".
+        ref_objects: dict of objects referenced by the target function, key is the name and value is the object.
+        classes_to_generate: list of classes which are defined in the same module as the target function.
+            Code generation is required for these classes.
+    """
+    func_module_name = func.__module__
+    if isinstance(func, FunctionType):
+        get_func_references(func, ref_objects)
+        to_import.update(
+            extract_submodule_imports(
+                func, [v for v in ref_objects.values() if isinstance(v, ModuleType)]
+            )
+        )
+        for v in [
+            v
+            for v in ref_objects.values()
+            if inspect.isclass(v) and v.__module__ == func_module_name
+        ]:
```

Comment:
Let's extract the list to a variable so it is easier to read?

Reply:
updated